### PR TITLE
Feature/refactor pdf converter

### PR DESCRIPTION
**Pull request by AKallala**
This PR refactors the Pdf converter view model and screen:

- It makes the generated pdf document be saved to the documents folder of the device
- It adds a function to handle the generation of the pdf document for all options
- it introduces a helper function to handle the saving process
- it introduces `PdfGenerationState` to give back feedback to the ui on the generation state of the file
- it adds an abort function that aborts the generation process if it's taking too long
- the screen also displays a loading indicator with a button that allows the user to abort the pdf generation process
- the scanner option is replaced by the doc to pdf option to allow the user to convert word, excel documents to pdf

### CODE REVIEW

**Review by MathieuxC — APPROVED** [review]
LGTM

File: app/src/main/java/com/github/se/eduverse/repository/PdfRepository.kt [hunk 1]
```diff
@@ -0,0 +1,272 @@
...
+      pdfDocument.finishPage(page)
+
+      return pdfDocument
+    } catch (e: Exception) {
+      // If an exception occurs during the conversion process, log the error and throw the exception
+      Log.e("convertTextToPdf", "Text conversion failed", e)
+      throw e
+    }
+  }
+
```
Comment (abj10) [inline app/src/main/java/com/github/se/eduverse/repository/PdfRepository.kt:185]: The PDF may not adhere to standard dimensions, which might be problematic for printing.

File: app/src/main/java/com/github/se/eduverse/viewmodel/PdfConverterViewModel.kt [hunk 2]
```diff
@@ -1,252 +1,145 @@
...
+            val pdfFile: PdfDocument =
+                when (converterOption) {
+                  PdfConverterOption.IMAGE_TO_PDF -> pdfRepository.convertImageToPdf(uri, context)
+                  PdfConverterOption.TEXT_TO_PDF -> pdfRepository.convertTextToPdf(uri, context)
+                  PdfConverterOption.DOCUMENT_TO_PDF -> throw Exception("Not implemented")
+                  PdfConverterOption.SUMMARIZE_FILE -> throw Exception("Not implemented")
+                  PdfConverterOption.EXTRACT_TEXT -> throw Exception("Not implemented")
+                  PdfConverterOption.NONE ->
+                      throw Exception("No converter option selected") // Should never happen
                 }
```
Comment (abj10) [inline app/src/main/java/com/github/se/eduverse/viewmodel/PdfConverterViewModel.kt:94]: Consider adding a loading indicator or a UI state reflecting the "InProgress" state to provide feedback during PDF generation.

**Review by abj10 — APPROVED** [review]
# Review: Feature/refactor pdf converter

The refactoring significantly improves the PDF conversion feature by organizing responsibilities across repositories, view models, and UI components. Here are a few detailed observations and suggestions for improvement:

1. **PdfRepository Improvements**
   - The addition of `PdfRepository` to handle PDF-related logic outside of the ViewModel is excellent. This modular approach enhances maintainability. To improve robustness, consider implementing more granular error handling in methods like `convertImageToPdf` and `convertTextToPdf`. Differentiating between file access issues and bitmap decoding errors could provide more specific debugging information.

2. **Enhanced User Feedback in UI**
   - The `LoadingIndicator` component with an abort feature is a thoughtful addition that enhances the user experience during lengthy tasks. It might be beneficial to verify that this indicator is displayed consistently across all PDF generation actions, ensuring users receive clear feedback for long operations.

3. **Consistent Use of ViewModel Factory**
   - Using a factory pattern to initialize `PdfConverterViewModel` with `PdfConverterScreen` is a solid choice for dependency injection, which supports maintainability and testability. Great job applying this pattern!

4. **UI Components Organization**
   - Refactoring UI elements like `OptionCard` contributes to clarity and modularity in the UI logic. Additionally, the addition of testing tags (such as `testTag("abortButton")`) will facilitate UI testing. For consistency, consider applying test tags to all user-facing buttons and interactive elements.

5. **Opportunities for Code Duplication Reduction**
   - In `PdfRepositoryImpl`, there are repetitive patterns in `convertImageToPdf` and `convertTextToPdf` regarding file processing and error handling. A helper function for shared tasks, such as file validation and stream handling, could reduce code redundancy and enhance readability.

File: app/src/main/java/com/github/se/eduverse/viewmodel/PdfConverterViewModel.kt (hunk 2 above)
Comment (AKallala) [reply app/src/main/java/com/github/se/eduverse/viewmodel/PdfConverterViewModel.kt:94]: It's already done

**Review by ismailIBNMAJAH — APPROVED** [review]
lgtm

**Pull request merged**